File: app/src/main/java/home/oleg/popularmovies/presentation/list/rvadapter/MovieAdapter.java
package home.oleg.popularmovies.presentation.list.rvadapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import home.oleg.popularmovies.R;
import home.oleg.popularmovies.data.network.NetworkMovies;
import home.oleg.popularmovies.presentation.model.MovieViewModel;

/**
 * Created by Oleg on 15.04.2017.
 */

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.ViewHolder> {

    private ArrayList<MovieViewModel> items;
    private final MovieAdapterOnClickHandler clickHandler;

    public ArrayList<MovieViewModel> getItems() {
        return items;
    }

    public interface MovieAdapterOnClickHandler {
        void onClick(MovieViewModel movieModel);
    }

    public MovieAdapter(MovieAdapterOnClickHandler clickHandler) {
        this.clickHandler = clickHandler;
        this.items = new ArrayList<>();
    }

    public MovieAdapter(ArrayList<MovieViewModel> items, MovieAdapterOnClickHandler clickHandler) {
        this(clickHandler);
        this.items = items;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.rv_movie_item, parent, false);
        return new ViewHolder(v, clickHandler);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(items.get(position));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void setItems(List<MovieViewModel> models) {
        if (!items.containsAll(models)) {
            items.clear();
            items.addAll(models);
            notifyDataSetChanged();
        }
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private Context context;
        private ImageView poster;
        private MovieAdapterOnClickHandler clickHandler;

        ViewHolder(View itemView, @NonNull MovieAdapterOnClickHandler clickHandler) {
            super(itemView);
            this.clickHandler = clickHandler;
            context = itemView.getContext();
            poster = (ImageView) itemView.findViewById(R.id.iv_poster);
        }

        void bind(@NonNull MovieViewModel model) {
            Picasso.with(context).load(NetworkMovies.IMAGE_URL + model.getImagePath()).into(poster);
            poster.setOnClickListener(view -> clickHandler.onClick(model));
        }
    }

}

File: app/src/main/java/home/oleg/popularmovies/presentation/list/rvadapter/MovieAdapter.kt
package home.oleg.popularmovies.presentation.list.rvadapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView

import com.squareup.picasso.Picasso
import home.oleg.popularmovies.BuildConfig

import home.oleg.popularmovies.R
import home.oleg.popularmovies.presentation.model.MovieViewModel
import kotlinx.android.synthetic.main.rv_movie_item.view.*
import javax.inject.Inject

/**
 * Created by Oleg on 15.04.2017.
 */

class MovieAdapter @Inject constructor() : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {

    var items: List<MovieViewModel> = emptyList()
    var clickListener: ((MovieViewModel) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieAdapter.ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.rv_movie_item, parent, false)
        return MovieAdapter.ViewHolder(v, clickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun getItemId(position: Int): Long {
        return items[position].id
    }

    fun setMovies(models: List<MovieViewModel>) {
        this.items = models
        notifyDataSetChanged()
    }

    class ViewHolder(itemView: View, private val clickListener: ((MovieViewModel) -> Unit)?) : RecyclerView.ViewHolder(itemView) {

        fun bind(model: MovieViewModel) = with(itemView) {
            Picasso.with(context).load(BuildConfig.BASE_IMAGE_URL + model.w185ImagePath).into(ivPoster)
            ivPoster.setOnClickListener { clickListener?.invoke(model) }
        }
    }

}
